Add "2ha.sh" script, managing 2-node Canonical K8s HA AA clusters (#692)
canonical/k8s-snap: docs/src/snap/howto/2-node-ha.md

> # 2-Node Active-Active High-Availability using Dqlite

Author: Whops, sure.

> # 2-Node Active-Active High-Availability using Dqlite

Author: I guess we can just say that ``Both nodes will be active members of the cluster, sharing the Kubernetes load.``

Kubernetes deployments, usually implying three or more nodes.

2-node HA clusters are sometimes preferred for cost savings and operational
efficiency considerations. Follow this guide to learn how Canonical Kubernetes
can achieve high availability with just two nodes while using the default
datastore, Dqlite.

Dqlite cannot achieve Raft quorum with less than three nodes. This means that
Dqlite will not be able to replicate data and the secondaries will simply
forward the queries to the primary node.

In the event of a node failure, database recovery will require following the
steps in the [Dqlite recovery guide].

## Proposed solution

Since Dqlite data replication is not available in this situation, we propose
using synchronous block level replication through
[Distributed Replicated Block Device] (DRBD).

The cluster monitoring and failover process will be handled by Pacemaker and
Corosync. After a node failure, the DRBD volume will be mounted on the standby
node, allowing access to the latest Dqlite database version.

Additional recovery steps are automated and invoked through Pacemaker.

## Alternatives

> Another possible approach is to use PostgreSQL with Kine and logical

Author: Right, so PostgreSQL was mentioned as an alternative in Antonio's document and it seemed like a good idea to cover it here as well. To be honest, I didn't get to try it out but I assumed it would be possible based on the Canonical Kubernetes docs on external datastores: https://documentation.ubuntu.com/canonical-kubernetes/latest/snap/howto/external-datastore/

I can remove the "alternatives" section for now.

replication. However, it is outside the scope of this document.

See the [external datastore guide] for more information on how Canonical
Kubernetes can be configured to use other datastores.

## Guide

### Prerequisites

* Ensure both nodes are part of the Kubernetes cluster.
  See the [getting started] and [add/remove nodes] guides.
* The user associated with the HA service has SSH access to the peer node and
  passwordless sudo configured. For simplicity, the default "ubuntu" user can
  be used.
* We recommend using static IP configuration.

> The [2ha.sh script] automates most operations related to the 2-node HA scenario

Author: Sure, I'd go with ``two-node-ha.sh``, it's easier to type when calling it.

and is included in the snap.

The first step is to install the required packages:

```
/snap/k8s/current/k8s/hack/2ha.sh install_packages
```

### DRBD

> For the purpose of this guide, we are going to use a loopback device as DRBD
> backing storage:

Reviewer: ```suggestion
This example uses a loopback device as DRBD
backing storage:
```

```
sudo dd if=/dev/zero of=/opt/drbd0-backstore bs=1M count=2000
```

Ensure that the loopback device is attached at boot time, before Pacemaker
starts.

```
cat <<EOF | sudo tee /etc/rc.local
#!/bin/sh
mknod /dev/lodrbd b 7 200
losetup /dev/lodrbd /opt/drbd0-backstore
EOF

sudo chmod +x /etc/rc.local

cat <<EOF | sudo tee /etc/systemd/system/rc-local.service
# This unit gets pulled automatically into multi-user.target by
# systemd-rc-local-generator if /etc/rc.local is executable.
[Unit]
Description=/etc/rc.local Compatibility
Documentation=man:systemd-rc-local-generator(8)
ConditionFileIsExecutable=/etc/rc.local
After=network.target

[Service]
Type=forking
ExecStart=/etc/rc.local start
TimeoutSec=0
RemainAfterExit=yes
GuessMainPID=no

[Install]
WantedBy=multi-user.target
EOF

sudo systemctl enable rc-local.service
sudo systemctl start rc-local.service
```

Let's configure the DRBD block device that will hold the Dqlite data.

> Ensure the correct node addresses are used.

Reviewer: ```suggestion
Ensure the correct node addresses are used.
```
This is important, instead of repeating it later make it stand out by putting it in a note box. Example for a note: https://github.com/canonical/k8s-snap/blob/c74b9f580f92b2fc6b77bf4ad692eb348b62a17f/docs/src/charm/howto/charm.md?plain=1#L18-L21

Author: Those are just some reminders to fill in the addresses instead of copy/pasting the following by mistake:

```
HAONE_ADDR=<firstNodeAddress>
HATWO_ADDR=<secondNodeAddress>
```

Do we need a note box for that?

```
# Disable the DRBD service, it will be managed through Pacemaker.
sudo systemctl disable drbd

HAONE_ADDR=<firstNodeAddress>
HATWO_ADDR=<secondNodeAddress>

cat <<EOF | sudo tee /etc/drbd.d/r0.res
resource r0 {
  on haone {
    device /dev/drbd0;
    disk /dev/lodrbd;
    address ${HAONE_ADDR}:7788;
    meta-disk internal;
  }
  on hatwo {
    device /dev/drbd0;
    disk /dev/lodrbd;
    address ${HATWO_ADDR}:7788;
    meta-disk internal;
  }
}
EOF

sudo drbdadm create-md r0
sudo drbdadm status
```

Let's create a mount point for the DRBD block device. Non-default mount points
need to be passed to the ``2ha.sh`` script mentioned above, see the script for
the full list of configurable parameters.

```
DRBD_MOUNT_DIR=/mnt/drbd0
sudo mkdir -p $DRBD_MOUNT_DIR
```

Run the following *once* to initialize the filesystem.

```
sudo drbdadm up r0

sudo drbdadm -- --overwrite-data-of-peer primary r0/0
sudo mkfs.ext4 /dev/drbd0

sudo drbdadm down r0
```

> Add the drbd device to the ``multipathd`` blacklist, ensuring that the multipath

Reviewer: ```suggestion
Add the DRBD device to the ``multipathd`` blacklist, ensuring that the multipath
```
Correct me if I am wrong here, I believe this would be capitalized as we have done so in other places.

service will not attempt to manage this device:

```
sudo cat <<EOF | sudo tee -a /etc/multipath.conf
blacklist {
    devnode "^drbd*"
}
EOF

sudo systemctl restart multipathd
```

### Corosync

> Let's prepare the Corosync configuration. Again, make sure to use the correct
> addresses.

Reviewer: ```suggestion
Prepare the Corosync configuration.
```

Author: I'd keep the reminder to update the addresses so that the readers don't copy/paste the following by mistake:

```
HAONE_ADDR=<firstNodeAddress>
HATWO_ADDR=<secondNodeAddress>
```

```
HAONE_ADDR=<firstNodeAddress>
HATWO_ADDR=<secondNodeAddress>

sudo mv /etc/corosync/corosync.conf /etc/corosync/corosync.conf.orig

cat <<EOF | sudo tee /etc/corosync/corosync.conf
totem {
  version: 2
  cluster_name: ha
  secauth: off
  transport:udpu
  interface {
    ringnumber: 0
    bindnetaddr: ${netaddr}
    broadcast: yes
    mcastport: 5405
  }
}

nodelist {
  node {
    ring0_addr: ${HAONE_ADDR}
    name: haone
    nodeid: 1
  }
  node {
    ring0_addr: ${HATWO_ADDR}
    name: hatwo
    nodeid: 2
  }
}

quorum {
  provider: corosync_votequorum
  two_node: 1
  wait_for_all: 1
  last_man_standing: 1
  auto_tie_breaker: 0
}
EOF
```

Follow the above steps on both nodes before moving forward.

### Pacemaker

Let's define a Pacemaker resource for the DRBD block device, which
ensures that the block device will be mounted on the replica in case of a
primary node failure.

[Pacemaker fencing] (stonith) configuration is environment specific and thus

> outside the scope of this guide. However, we highly recommend using fencing
> if possible to reduce the risk of cluster split-brain situations.

Reviewer: ```suggestion
outside the scope of this guide. Using fencing is highly recommended
if it is possible to reduce the risk of cluster split-brain situations.
```

property stonith-enabled=false
property no-quorum-policy=ignore
primitive drbd_res ocf:linbit:drbd params drbd_resource=r0 op monitor interval=29s role=Master op monitor interval=31s role=Slave
ms drbd_master_slave drbd_res meta master-max=1 master-node-max=1 clone-max=2 clone-node-max=1 notify=true
primitive fs_res ocf:heartbeat:Filesystem params device=/dev/drbd0 directory=${DRBD_MOUNT_DIR} fstype=ext4
colocation fs_drbd_colo INFINITY: fs_res drbd_master_slave:Master
order fs_after_drbd mandatory: drbd_master_slave:promote fs_res:start
commit
show
quit
EOF
```

Before moving forward, let's ensure that the DRBD Pacemaker resource runs on
the primary (voter) Dqlite node.

In this setup, only the primary node holds the latest Dqlite data, which will
be transferred to the DRBD device once the clustered service starts.
This is automatically handled by the ``2ha_k8s.sh start_service`` command.

```
sudo k8s status
sudo drbadadm status
sudo crm status
```

If the DRBD device is assigned to the secondary Dqlite node (spare), move it
to the primary like so:

```
sudo crm resource move fs_res <primary_node_name>

# remove the node constraint.
sudo crm resource clear fs_res
```

> ### Kubernetes services
> 
> We can now turn our attention to the Kubernetes services. Ensure that the k8s
> snap services no longer start automatically. Instead, they will be managed by a
> wrapper service.
> 

Reviewer: ```suggestion
### Managing Kubernetes Snap Services

For the two-node HA setup k8s snap services should no longer start automatically. Instead, they will be managed by a wrapper service.

```

```
for f in `sudo snap services k8s  | awk 'NR>1 {print $1}'`; do
    echo "disabling snap.$f"
    sudo systemctl disable "snap.$f";
done
```

> The next step is to define the wrapper service. Add the following to

Reviewer: You can consider making this a sub-header.

``/etc/systemd/system/2ha_k8s.service``. Note that the sample uses the
``ubuntu`` user, feel free to use a different one as long as the prerequisites

> are met.

Reviewer: Consider making this a note.

```
[Unit]
Description=K8s service wrapper handling Dqlite recovery for 2-node HA setups.
After=network.target pacemaker.service

[Service]
User=ubuntu
Group=ubuntu
Type=oneshot
ExecStart=/bin/bash /snap/k8s/current/k8s/hack/2ha.sh start_service
ExecStop=/bin/bash sudo snap stop k8s
RemainAfterExit=true

[Install]
WantedBy=multi-user.target
```

```{note}
The ``2ha.sh start_service`` command used by the service wrapper automatically
detects the expected Dqlite role based on the DRBD state and takes the
necessary steps to bootstrap the Dqlite state directories, synchronize with the

> peer node (if available) and recover the database.

Reviewer: Break up this sentence in two please.

```

> We need the ``2ha_k8s`` service to be restarted once a DRBD failover occurs.

Reviewer: ```suggestion
When a DRBD failover occurs the ``2ha_k8s`` service needs to be restarted.
```

> For that, we are going to define a separate service that will be invoked by

Reviewer: ```suggestion
To accomplish this, define a separate service that will be invoked by
```

Pacemaker. Create a file called
``/etc/systemd/system/2ha_k8s_failover.service`` containing the following:

```
[Unit]
Description=Managed by Pacemaker, restarts 2ha_k8s on failover.
After=network.target home-ubuntu-workspace.mount

[Service]
Type=oneshot
ExecStart=systemctl restart 2ha_k8s
RemainAfterExit=true
```

Reload the systemd configuration and set ``2ha_k8s`` to start automatically.
Notice that ``2ha_k8s_failover`` must not be configured to start automatically,
but instead is going to be managed through Pacemaker.

```
sudo systemctl enable 2ha_k8s
sudo systemctl daemon-reload
```

Make sure that both nodes have been configured using the above steps before

> moving forward.

Reviewer: Please number the steps and then rephrase this to something like please do the following steps on both nodes before continuing with the setup.

Author: Should we number each step starting with ```### Prerequisites```?

> We can now define a new Pacemaker resource that will invoke the

Reviewer: ```suggestion
Define a new Pacemaker resource that will invoke the
```

Author: Right, so PostgreSQL was mentioned as an alternative in Antonio's document and it seemed like a good idea to cover it here as well. To be honest, I didn't get to try it out but I assumed it would be possible based on the Canonical Kubernetes docs on external datastores: https://documentation.ubuntu.com/canonical-kubernetes/latest/snap/howto/external-datastore/

``2ha_k8s_failover`` service when a DRBD failover occurs.

```
sudo crm configure <<EOF
primitive ha_k8s_failover_service systemd:2ha_k8s_failover op start interval=0 timeout=120 op stop interval=0 timeout=30
order failover_after_fs mandatory: fs_res:start ha_k8s_failover_service:start
colocation fs_failover_colo INFINITY: fs_res ha_k8s_failover_service
commit
show
quit
EOF
```

> The setup is ready, start the HA k8s service on both nodes:

Reviewer: ```suggestion
Once the setup is complete on both nodes, start the two-node HA k8s service on each node:
```

```
sudo systemctl start 2ha_k8s
```

## Troubleshooting

> ### Dqlite recovery failing because of unexpected data segments

Reviewer: I believe we need to capitalize more "important" words in our headers here and in other places.

Author: We could, although the other docs don't seem to do this:

* https://documentation.ubuntu.com/canonical-kubernetes/latest/snap/howto/external-datastore/

* https://documentation.ubuntu.com/canonical-kubernetes/latest/snap/howto/backup-restore/

Furthermore, it might affect readability.

> ### Dqlite recovery failing because of unexpected data segments

Reviewer: Add some words in between two headers please.

Dqlite recovery may fail if there are data segments past the latest snapshot.

```
Error: failed to recover k8s-dqlite, error: k8s-dqlite recovery failed, error:
recover failed with error code 1, error details: raft_recover(): io:
closed segment 0000000000002369-0000000000002655 is past last snapshot
snapshot-2-2048-642428, pre-recovery backup:
/var/snap/k8s/common/recovery-k8s-dqlite-2024-09-05T082644Z-pre-recovery.tar.gz
```

Remove the offending segments and restart the ``2ha_k8s`` service.

### DRBD split brain

> The DRBD cluster may enter a split brain state and stop synchronizing. The

Reviewer: Please add a reference to a split brain explanation.

> # 2-Node Active-Active HA using Dqlite

Reviewer 2: High-Availability

## Rationale

> High availability is a mandatory requirement for most production-grade

Reviewer 2: (HA)

using synchronous block level replication through DRBD.

The cluster monitoring and failover process will be handled by Pacemaker and
Corosync. In the event of a node failure, the DRBD volume will be mounted on
the replica, which can then access the most recent version of the Dqlite database.

Additional recovery steps are automated and invoked through Pacemaker.

## Alternatives

Another possible approach is to use PostgreSQL with Kine and logical replication.
However, it is outside the scope of this document.

See the [external datastore guide] for more information on how Canonical
Kubernetes can be configured to use other datastores.

## Guide

### Prerequisites

Make sure that:

* Both nodes have joined the Kubernetes cluster.
  See the [getting started] and [add/remove nodes] guides.
* The user associated with the HA service has SSH access to the peer node and
  passwordless sudo configured. For simplicity, the default "ubuntu" user can
  be used.
* We recommend using static IP configuration.

The [2ha.sh script] automates most operations related to the 2-node HA scenario.
Retrieve it like so:

```
sudo mkdir -p /var/snap/k8s/common

> repo=https://raw.githubusercontent.com/petrutlucian94/k8s-snap

Reviewer 2: This script should be part of the snap. If you put it into hack then it should either be available at `/snap/k8s/current/k8s/hack` or you would need to update the `snapcraft.yaml` to add this directory to the snap.

Author: Great, I guess we can just use ``/snap/k8s/current/k8s/hack/2ha.sh``.

```

The first step is to install the required packages:

```
/var/snap/k8s/common/2ha.sh install_packages
```

### DRBD

For the purpose of this guide, we are going to use a loopback device as DRBD
backing storage:

```
sudo dd if=/dev/zero of=/opt/drbd0-backstore bs=1M count=2000
```

Ensure that the loopback device is attached at boot time, before Pacemaker
starts.

```
cat <<EOF | sudo tee /etc/rc.local
#!/bin/sh
mknod /dev/lodrbd b 7 200
losetup /dev/lodrbd /opt/drbd0-backstore
EOF

sudo chmod +x /etc/rc.local

cat <<EOF | sudo tee /etc/systemd/system/rc-local.service
# This unit gets pulled automatically into multi-user.target by
# systemd-rc-local-generator if /etc/rc.local is executable.
[Unit]
Description=/etc/rc.local Compatibility
Documentation=man:systemd-rc-local-generator(8)
ConditionFileIsExecutable=/etc/rc.local
After=network.target

[Service]
Type=forking
ExecStart=/etc/rc.local start
TimeoutSec=0
RemainAfterExit=yes
GuessMainPID=no

[Install]
WantedBy=multi-user.target
EOF

sudo systemctl enable rc-local.service
sudo systemctl start rc-local.service
```

Let's configure the DRBD block device that will hold the Dqlite data.

> Make sure to use the right node addresses.

Reviewer 2: Ensure the correct node addresses are used.

```
# Disable the DRBD service, it will be managed through Pacemaker.
sudo systemctl disable drbd

HAONE_ADDR=<firstNodeAddress>
HATWO_ADDR=<secondNodeAddress>

cat <<EOF | sudo tee /etc/drbd.d/r0.res
resource r0 {
  on haone {
    device /dev/drbd0;
    disk /dev/lodrbd;
    address ${HAONE_ADDR}:7788;
    meta-disk internal;
  }
  on hatwo {
    device /dev/drbd0;
    disk /dev/lodrbd;
    address ${HATWO_ADDR}:7788;
    meta-disk internal;
  }
}
EOF

sudo drbdadm create-md r0
sudo drbdadm status
```

Let's create a mount point for the DRBD block device. Non-default mount points

> need to be passed to the 2ha.sh script mentioned above, see the script for the

Reviewer 2: put `2ha.sh` into ticks `

full list of configurable parameters.

```
DRBD_MOUNT_DIR=/mnt/drbd0
sudo mkdir -p $DRBD_MOUNT_DIR
```

> Run the following *once* to initialize the filesystem.

Reviewer 2: Maybe add what would happen if you accidentally execute it multiple times?
e.g. "Executing this multiple times would cause ..."

Author: Not much since we haven't yet stored any data yet, not sure if it's worth mentioning it.

The main reason why I emphasised "once" is that below there's a line saying ``Follow the above steps on both nodes before moving forward.``, however this step in particular doesn't have to be executed twice.

service will not attempt to manage this device:

```
sudo cat <<EOF | sudo tee -a /etc/multipath.conf
blacklist {
    devnode "^drbd*"
}
EOF

sudo systemctl restart multipathd
```

### Corosync

Let's prepare the Corosync configuration. Again, make sure to use the correct
addresses.

```
HAONE_ADDR=<firstNodeAddress>
HATWO_ADDR=<secondNodeAddress>

sudo mv /etc/corosync/corosync.conf /etc/corosync/corosync.conf.orig

cat <<EOF | sudo tee /etc/corosync/corosync.conf
totem {
  version: 2
  cluster_name: ha
  secauth: off
  transport:udpu
  interface {
    ringnumber: 0
    bindnetaddr: ${netaddr}
    broadcast: yes
    mcastport: 5405
  }
}

nodelist {
  node {
    ring0_addr: ${HAONE_ADDR}
    name: haone
    nodeid: 1
  }
  node {
    ring0_addr: ${HATWO_ADDR}
    name: hatwo
    nodeid: 2
  }
}

quorum {
  provider: corosync_votequorum
  two_node: 1
  wait_for_all: 1
  last_man_standing: 1
  auto_tie_breaker: 0
}
EOF
```

Follow the above steps on both nodes before moving forward.

### Pacemaker

Let's define a Pacemaker resource for the DRBD block device, which
ensures that the block device will be mounted on the replica in case of a
primary node failure.

> Pacemaker fencing (stonith) configuration is environment specific and thus

Reviewer 2: nit: link to fencing/stonith explanation?

outside the scope of this guide. However, we highly recommend using fencing
if possible to reduce the risk of cluster split-brain situations.

```
HAONE_ADDR=<firstNodeAddress>
HATWO_ADDR=<secondNodeAddress>
DRBD_MOUNT_DIR=${DRBD_MOUNT_DIR:-"/mnt/drbd0"}

sudo crm configure <<EOF
property stonith-enabled=false
property no-quorum-policy=ignore
primitive drbd_res ocf:linbit:drbd params drbd_resource=r0 op monitor interval=29s role=Master op monitor interval=31s role=Slave
ms drbd_master_slave drbd_res meta master-max=1 master-node-max=1 clone-max=2 clone-node-max=1 notify=true
primitive fs_res ocf:heartbeat:Filesystem params device=/dev/drbd0 directory=${DRBD_MOUNT_DIR} fstype=ext4
colocation fs_drbd_colo INFINITY: fs_res drbd_master_slave:Master
order fs_after_drbd mandatory: drbd_master_slave:promote fs_res:start
commit
show
quit
EOF
```

Before moving forward, let's ensure that the DRBD Pacemaker resource runs on
the primary (voter) Dqlite node.

> Remember that in our case only the primary node contains the latest Dqlite data,

Reviewer 2: In this setup, only the primary node holds the latest Dqlite data, which will be transferred to the DRBD device

which will be transfered to the DRBD device once the clustered service starts.
This is automatically handled by the ``2ha_k8s.sh start_service`` command.

```
sudo k8s status
sudo drbadadm status
sudo crm status
```

If the DRBD device is assigned to the secondary Dqlite node (spare), move it
to the primary like so:

```
sudo crm resource move fs_res <primary_node_name>

# remove the node constraint.
sudo crm resource clear fs_res
```

### Kubernetes services

We can now turn our attention to the Kubernetes services. Ensure that the k8s

> snap services no longer start automatically. Instead, they will be manged by a

Reviewer 2: managed
